--- format.mli	2017-11-04 21:05:25.000000000 +0000
+++ format_.mli	2017-12-17 20:57:07.000000000 +0000
@@ -177,6 +177,20 @@
   has just been split. Otherwise, ignore the next formatting
   command. *)
 
+val print_string_if_newline : string -> unit
+(** Similar to [print_if_newline] followed by [print_string] except that the
+  length of the string does not contribute to the width of the enclosing
+  box. *)
+
+val print_break_or_string_if_newline : int -> int -> string -> unit
+(** Print a full break hint if the preceding line has not just been split.
+  Otherwise, print a string. *)
+
+val print_fits_or_breaks : string -> int -> int -> string -> unit
+(** [print_fits_or_breaks fits nspaces offset breaks] prints [fits] if the
+  enclosing box fits on one line. Otherwise, prints a break as per
+  [print_break nspaces offset] followed by [breaks]. *)
+
 (** {6 Margin} *)
 
 val set_margin : int -> unit
@@ -526,6 +540,11 @@
 val pp_print_flush : formatter -> unit -> unit
 val pp_print_newline : formatter -> unit -> unit
 val pp_print_if_newline : formatter -> unit -> unit
+val pp_print_string_if_newline : formatter -> string -> unit
+val pp_print_break_or_string_if_newline :
+  formatter -> int -> int -> string -> unit
+val pp_print_fits_or_breaks :
+  formatter -> string -> int -> int -> string -> unit
 val pp_set_tags : formatter -> bool -> unit
 val pp_set_print_tags : formatter -> bool -> unit
 val pp_set_mark_tags : formatter -> bool -> unit
